### PR TITLE
restrict volume names

### LINKED ISSUE #16132: Volumes: relative host-path creates named volumes
While reviewing a PR (https://github.com/docker/docker/pull/16100#issuecomment-138455740), I discovered that using a *relative* path as host path, will create a named volume, instead of erroring out or creating a bind-mounted volume.

To reproduce;

```bash
docker run -v ./:/foobar ubuntu
```

Then inspect the container;

    "Mounts": [
        {
            "Name": "./",
            "Source": "/var/lib/docker/volumes/_data",
            "Destination": "/foobar",
            "Driver": "local",
            "Mode": "z",
            "RW": true
        }
    ],

From the inspect, it's visible that a volume named `"./"` is created

Note that using `-v ~/:/foobar` *does work*;

    "Mounts": [
        {
            "Source": "/root",
            "Destination": "/foobar",
            "Mode": "",
            "RW": true
        }
    ],

**Expected result**

Docker producing an error

I think we should (at least, for now) only allow `[a-zA-Z0-9][a-zA-Z0-9_.-]+` for volume-names, anything else should produce an error.

**Actual result**

Docker created a volume named `"./"`

docker version
```
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   8562c42
 Built:        Sat Sep  5 12:34:19 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   8562c42
 Built:        Sat Sep  5 12:34:19 UTC 2015
 OS/Arch:      linux/amd64
```

docker info
```
Containers: 16
Images: 579
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 620
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-61-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 1
Total Memory: 994 MiB
Name: [REDACTED]
ID: [REDACTED]
```

### PR DESCRIPTION
**Pull request by xlgao-zju**
Closes #16132 

```
Agree if [a-zA-Z0-9][a-zA-Z0-9_.-]+ is what we want 
```

at first.

And if the definitive Regex gets changed, I'll update this pr.
@thaJeztah

### CODE REVIEW

**Comment by duglin** [comment]
please merge your commits.

please put `Closes: #16132` in your first comment so that git closes the issue when this PR is merged.

ping @cpuguy83 since this is about volumes.

**Comment by cpuguy83** [comment]
I'm so bad with regexes I'm not going to comment much on that... other than +1 for a regex since it's weird to support relative paths.

**Comment by thaJeztah** [comment]
> I'm so bad with regexes I'm not going to comment much on that... other than +1 for a regex since it's weird to support relative paths.

There's an online tester here; https://regex-golang.appspot.com/assets/html/index.html if you want to check what's accepted and not.

To explain this Regex, it currently accepts volume that;

- Start with an alphanumeric character (a-z, A-Z, 0-9), followed by
- At *least* one alphanumeric character (a-z, A-Z, 0-9), dot, underscore or dash

This means that *single character* volume names (e.g. `A`) are *not* accepted. I was thinking if we should, but I think it's better *not* to, to prevent `docker run -v C:\Users\foo\` leading to a named volume `C`.

There's one concern I have; we accept both uppercase and lowercase characters. Given that the Daemon keeps reference to volumes by creating a directory in `/var/lib/docker/volumes/<volumename>`, this will cause problems if the daemon is running on a case-insensitive filesystem (read: `docker daemon` on Windows). On such system, `MyVolume` and `myvolume` will be the *same* volume, whereas on Linux, they will be *different* volumes.

I don't think the Windows  daemon already supports this feature, but it may be added in some future, so we may need to think of that *now*.

TL;DR; if we want to be "safe" we should only support lowercased names, and change the regex to 

`^[a-z0-9][a-z0-9_.-]+$`

**Comment by thaJeztah** [comment]
ping @jhowardmsft @swernli @ahmetalpbalkan for input on the "Windows" side of this ^^

**Comment by jhowardmsft** [comment]
@jstarks I'm actually working on volume support in Docker for Windows right now. Obviously this change is in a _unix file right now and doesn't directly impact Windows. 

On Windows we need to support any/all of (in case insensitive)
c:\foo
c:
\foo

So docker run -v c:\foo:d:\bar:wr would be an option. Obviously that is almost impossible to deterministically parse due to the use of colon being both the separator of the parameter and used as part of the path. The prototype I'm working on addresses this through the use of square brackets where a path with a colon is required. [c:\foo]:\bar:wr.

**Comment by thaJeztah** [comment]
@jhowardmsft I think _if we **require** a named volume to have a minimum length of two characters_, that the brackets won't be necessary; `C:` is a single character, so won't be interpreted as a "named volume". Some smart regex should be able to match the cases you mentioned;

I just tried it a bit, and came with this quick & dirty regex (not including unc-paths yet);

    ^((?P<volumename>[a-zA-Z0-9][a-zA-Z0-9_.-]+)|(?P<path1>[a-zA-Z]:(\\[^\:\\]+)*\\?))(:(?P<path2>[a-zA-Z]:(\\[^\:\\]+)*\\?)(:(?P<readwrite>ro|rw))?(,(?P<relabel>z|Z))?)?$

You can play around with that on https://regex-golang.appspot.com/assets/html/index.html. This gist breaks the regex down a bit: https://gist.github.com/thaJeztah/6185659e4978789fb2b2, and includes some example values for testing.

@jhowardmsft what are your thoughts on "case sensitivity" for *named* volumes on Windows? Will Window be able to have two volumes, one named `Foobar` and one named `foobar`? or should they be the same volume?

**Comment by xlgao-zju** [comment]
uh, I think I should wait for your definitive Regex...

**Comment by thaJeztah** [comment]
Lets ignore the Regex for Windows for now; the long Regex above is just a POC, not relevant to this PR.

For now we need to define "accepted" volume-names for Linux, but keep in mind that those names should not
cause problems once we introduce named volumes on Windows.

My thoughts are:

1. Minimum length 2 characters (to disallow creating a volume named `c`, which cannot be used on Windows as `c:` is a drive-letter)
2. Must start with an alphanumeric character (no `.`, `-`; we don't want volumes named `--foo` or `.hidden`)
3. Only allow alphanumeric characters, `_`, `.` and `-`. No whitespace. 
4. Either only allow *lowercase* characters, <br />
    **OR**
5. Treat volume-names case-insensitive and normalize them (both `--name MyVolume` and `--name mYvolUme` would create a volume named `myvolume`)

Both `4.` and `5.` make sure that names behave the same on Windows and Linux. Out of those two, my preference goes to `4.` as it's more explicit, and easier to "loosen up" in a later stage (if possible). Also changing upper to lowercase can be non-trivial if Unicode characters are accepted in the future.


ping @cpuguy83 @jhowardmsft @calavera can we agree on the above? which out of `4.` and `5.` would have your preference?

**Comment by cpuguy83** [comment]
Also see `$ sudo docker run -it --volume-driver sshfs --volume root@1.2.3.4#/data:/data busybox sh`
From https://github.com/vieux/docker-volume-sshfs

**Comment by jhowardmsft** [comment]
@thaJeztah Re `4` or `5` above - I don't think it matters at all from a Windows perspective. This is validation being done in _unix.go files and is therefore not relevant at all for Windows. The Windows equivalent of parsing the (bind) mount path in platform semantics will be quite different.

**Comment by calavera** [comment]
The regexp is fine, I would have used `^\w[\w_.-]+$`, but it basically means the same.

There are no tests though :pensive:

**Comment by thaJeztah** [comment]
@calavera I decided to not use `\w` here, because a-z was more explicit (and IIRC, used in other places), I'm good changing it though.

@jhowardmsft Yes, this is Linux only now; but I wanted to be sure we could give the same user experience on Windows (ie, being able to use the same *name* on Windows as on Linux) - not talking about bind-mounts here.

@cpuguy83 good one; I think we need to think of we still want to support that, or if `--opts` should be used in stead. @vieux any thoughts on that?

**Comment by jfrazelle** [comment]
can you make your commit message actually describe what is changing

**Comment by xlgao-zju** [comment]
Ok, I'll change it tomorrow.
@jfrazelle

**Comment by xlgao-zju** [comment]
So...what's the definitive Regex?
@thaJeztah

**Comment by thaJeztah** [comment]
I think the current one is good, but would love to hear from @cpuguy83 and @vieux about https://github.com/docker/docker/pull/16140#issuecomment-139302680

File: daemon/volumes_unix.go [hunk 1]
```diff
@@ -64,6 +65,7 @@ func parseBindMount(spec, volumeDriver string) (*mountPoint, error) {
 		RW: true,
 	}
 	arr := strings.Split(spec, ":")
+	reg := regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9_.-]+$`)
```
Comment (jhowardmsft) [inline daemon/volumes_unix.go]: I think it would be better if this were declared immediately above where it is used

Comment (calavera) [inline daemon/volumes_unix.go]: agreed

**Comment by xlgao-zju** [comment]
Code is updated...and I'm still waiting for the definitive Regex...

**Comment by calavera** [comment]
Now that I think about it, and after playing with the volume cli, I think we should not restrict the names with a regexp like this one. I think we should only check if the name starts with `./`, with `../` or contain a `:`. Other character restrictions are completely unnecessary.

**Comment by jhowardmsft** [comment]
I'm (perhaps controversially) going to say at least from a Windows perspective, having "sensible" names makes the horrendous regex for parsing the spec (due to : being part of a Windows path) MUCH easier. (I'm already using this same regex in the volume branch I'm working on for Windows support to determine if the first element is a name or host directory). At least as Windows does not have any down-level baggage to deal with, it makes sense limiting the name in the volume cli as well on Windows.

**Comment by thaJeztah** [comment]
> I think we should only check if the name starts with ./, with ../ or contain a :. Other character restrictions are completely unnecessary.

It seems to work, but, not sure if e.g. `docker run -v foo/../../bar/:/foo ...` is something we want

For fun, I created a nice looking volume:
image

**Comment by jfrazelle** [comment]
that worked lol?

**Comment by thaJeztah** [comment]
Yes! Thought it was cute (although my fish completion started to freak out afterwards, for a bit)

**Comment by xlgao-zju** [comment]
The id of container in runc is following those code.
https://github.com/opencontainers/runc/blob/master/libcontainer/factory_linux.go#29
https://github.com/opencontainers/runc/blob/master/libcontainer/factory_linux.go#261
Shall we do this like that?

**Comment by calavera** [comment]
I really really don't want to stop people from typing the characters they want. This is what we need to fix:

If the name looks like a relative path, based in the rules the OS applies to consider something a relative path, FAIL, otherwise PASS.

Restricting the name to only a subset of "valid" characters is not the valid answer.

**Comment by xlgao-zju** [comment]
So, only when the name starts with `./` and `../`, FAIL?

**Comment by calavera** [comment]
that's good enough to me. Can you declare the error in the new `errors package`? I think it could be more helpful if the message said something like `this looks like a relative path, but it's taken as a name`

**Comment by xlgao-zju** [comment]
Ok, I'll do it.

**Comment by xlgao-zju** [comment]
code is updated, please review.
@calavera

**Comment by xlgao-zju** [comment]
don't know why FAIL on windows at `docker_cli_exec_test.go:616: DockerSuite.TestExecStartFails`

**Comment by thaJeztah** [comment]
@xlgao-zju you can ignore that one, there's a problem with that test.

**Comment by xlgao-zju** [comment]
@thaJeztah So, could you review the code for me?

**Comment by thaJeztah** [comment]
@xlgao-zju I'm not really a code reviewer, but for what I can see, it looks good, and implements what @calavera suggested in https://github.com/docker/docker/pull/16140#issuecomment-141156566

**Comment by vdemeester** [comment]
Not sure if we need an integration test or not as it is covered by the unit test, but other thab that, based on previous discussion : LGTM

**Comment by xlgao-zju** [comment]
@thaJeztah thx.
And ping @calavera for code review.

**Comment by vdemeester** [comment]
@calavera is in vacation 👻 so ping @runcom @cpuguy83 @duglin @LK4D4

**Comment by thaJeztah** [comment]
t.b.h, I'm still concerned with allowing *any* character in these names; can we guarantee that this works on any system? Given that (e.g.) multi byte characters can be used (see my earlier example https://github.com/docker/docker/pull/16140#issuecomment-141254127), this will implicitly add extra requirements to the host that docker is running on (charsets can be a real PITA; https://github.com/docker/docker/issues/16516#issuecomment-142640574). Will it work in combination with Swarm (not sure that has been tested yet)?

My earlier "floppy disk" example also broke my Fish completion, which didn't handle the multi-byte character properly, causing all completions to act erratically.

TL;DR I'm all for not restricting users from using any name they like, just concerned that we may end up having to add restrictions in future, resulting in backwards incompatible changes (similar to https://github.com/docker/distribution/issues/1062)

/cc @docker/maintainers

**Comment by jhowardmsft** [comment]
Where are we going with this? It seems to have stalled. I relaxed the Windows volume naming in #16433 at the request of @calavera to only allow NTFS permissible characters, and exclude NTFS blocked device names. I would like to keep parity as far as possible with Unix.

**Comment by tianon** [comment]
Sorry if I missed it searching the thread, but why aren't we using https://golang.org/pkg/path/filepath/#IsAbs ?

**Comment by xlgao-zju** [comment]
Not restricting users from using any name they like is good. But I don't think users can get what they really want by using the name which starts with `./` or `../`.

**Comment by duglin** [comment]
Not trying to backtrack on this, but why are we looking to restrict this at all?
IOW, why not just take the string(path) and see if it points to a valid dir (or file, if files are allowed) and if so then mount it.  Why are we looking to restrict this at all beyond "point to something valid on the current filesystem" ?

**Comment by duglin** [comment]
Let me elaborate, if they are in `/root/mydir` and use a path of `../` we should convert to an abs path (`/root`) and then use that.

**Comment by jfrazelle** [comment]
I don't know if I agree with ../ why not give an abs path.. idk I need to think about it more






On Thu, Oct 8, 2015 at 8:44 AM -0700, "Doug Davis" <notifications@github.com> wrote:












Let me elaborate, if they are in /root/mydir and use a path of ../ we should convert to an abs path (/root) and then use that.

**Comment by cpuguy83** [comment]
Yeah, relative paths don't make sense here unless we start doing some remote filesystem magic.

**Comment by duglin** [comment]
as long as we can calculate the abs path does it really matter if they specify ../ or /root ?

**Comment by duglin** [comment]
oh right - this is the path on the host, not the CLI. doi - sorry.  I only work on the host machine so I keep forgetting about that  :-)

**Comment by cpuguy83** [comment]
@duglin What does a relative path even mean in this regard? Most docker clients are remote clients. This is even harrier when you throw swarm into the mix.

File: daemon/volumes_unix.go [hunk 2]
```diff
@@ -92,6 +92,11 @@ func parseBindMount(spec, volumeDriver string) (*mountPoint, error) {
 	}
 
 	if len(source) == 0 {
+		//validate the name of named volume
+		if strings.HasPrefix(name, "./") || strings.HasPrefix(name, "../") {
```
Comment (duglin) [inline daemon/volumes_unix.go]: Should we use go's path.IsAbs() instead to be safe?

File: errors/daemon.go [hunk 3]
```diff
@@ -385,6 +385,14 @@ var (
 		HTTPStatusCode: http.StatusInternalServerError,
 	})
 
+	// ErrorCodeVolumeName is generated when the name of named volume isn't valid.
+	ErrorCodeVolumeName = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "VOLUMENAME",
+		Message:        "%s looks like a relative path, but it's taken as a name. And it is not a valid name.",
+		Description:    "The name of named volume must not start with ./ or ../",
```
Comment (duglin) [inline errors/daemon.go]: How about:
```
The path to the named volume must be an absolute path. For example, it must not start with ./ or ../.
```

Comment (duglin) [inline errors/daemon.go]: deleting comment.

File: daemon/volumes_unix.go (hunk 2 above)
Comment (duglin) [inline daemon/volumes_unix.go]: erasing comment.

**Comment by xlgao-zju** [comment]
@duglin  I think we can't use `filepath.IsAbs`. Since the name of volume will always get `false` with `filepath.IsAbs`. 
Relative code:
https://github.com/docker/docker/blob/master/daemon/volumes_unix.go#L401

File: daemon/volumes_unix.go (hunk 2 above)
Comment (cpuguy83) [inline daemon/volumes_unix.go]: `.` and `..` are problems as well... and I suspect anything that looks like a path would cause some issues.

Comment (cpuguy83) [inline daemon/volumes_unix.go]: This seems to block out the bad stuff: `if regexp.MatchString("(^.+[^0-9A-Za-z_]+$)|(/)", name)`

**Comment by crosbymichael** [comment]
Sorry for coming in late but did we ever support relative volumes?  I didn't think so and we always required the source to be an ABS path.

**Comment by cpuguy83** [comment]
@crosbymichael No, and the title is a misnomer.
This is about dealing (or has become about) with bad values for volume names.

**Comment by crosbymichael** [comment]
Ok, i think anyone with write access can change the titles of PRs so that they are not confusing and misleading.  I would suggest doing so in cases like this. ;)

Anyways, i think you can treat anything with a leading `/` as a volume path, everything else should follow the container name regex.

**Comment by thaJeztah** [comment]
@crosbymichael that was my original proposal, however this would break https://github.com/docker/docker/pull/16140#issuecomment-139302680

And @calavera wants users to give more freedom in naming their volumes, which could in turn lead to things we may not want to support (or cannot guarantee will work); https://github.com/docker/docker/pull/16140#issuecomment-141254127

File: daemon/volumes_unix.go (hunk 2 above)
Comment (xlgao-zju) [inline daemon/volumes_unix.go]: it seems not good enough...
`.` can't be blocked...

Comment (cpuguy83) [inline daemon/volumes_unix.go]: ah crap, I must have changed without testing `.`...

**Comment by cpuguy83** [comment]
I'm going to say +1 on using the container name regexp. If it breaks people doing silly things then fine.
A name should be a name, not encoded options.
See https://github.com/docker/docker/issues/16132#issuecomment-138553918 from early September.

But If not, we need to disallow anything that looks like a file path since this for sure messes up the local volume driver, and probably would not do good things to most other drivers. That means anything that has a `/` in it. Probably anything that has `.` as the prefix since that's probably an accident for people looking for a relative path.. and for certain `.`, `..`, etc. since that is just out right broken in both the client and the local volume driver.

ping @calavera who had reservations about restricting this much.

Also @xlgao-zju thanks for putting up with going back and forth on this.
LMK if you'd like for someone to carry this or not.

**Comment by calavera** [comment]
> I'm going to say +1 on using the container name regexp. If it breaks people doing silly things then fine.
> A name should be a name, not encoded options.

using more than a-z0-9 characters is not doing silly things. We cannot restrict characters that the operating system accepts as valid for directory names only because we don't know how to differentiate absolute paths from a relative paths.

**Comment by crosbymichael** [comment]
@calavera checking for an abs path is easy in a cross platform way ;)

https://golang.org/pkg/path/filepath/#Abs

**Comment by tiborvass** [comment]
@calavera @crosbymichael unfortunately for windows the stdlib is not sufficient yet: we're using: https://github.com/docker/docker/blob/master/pkg/system/filesys_windows.go#L68

**Comment by calavera** [comment]
@crosbymichael I'm fully aware. But what started as "we should not allow what looks like relative paths as volume names" has turned into "if we let people use emojis as names we all are going to die" :smile_cat: :trollface:

**Comment by tiborvass** [comment]
Closing in favor of #17213

Thanks for your contribution @xlgao-zju ! Really appreciate it!

**Comment by xlgao-zju** [comment]
That's ok. I will carry on!

**Comment by thaJeztah** [comment]
Thanks @xlgao-zju! Sorry for all the bike shedding here :heart:

**Comment by xlgao-zju** [comment]
@thaJeztah That's all right. Working  with you guys is so happy.
